external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -438,55 +440,53 @@ private void commitOffsetsForAckedTuples() {
     // ======== Ack =======
     @Override
     public void ack(Object messageId) {
-        if (!isAtLeastOnce()) {
-            // Only need to keep track of acked tuples if commits are done based on acks
-            return;
-        }
-
+        // Only need to keep track of acked tuples if commits to Kafka are done after a tuple ack is received
         final KafkaSpoutMessageId msgId = (KafkaSpoutMessageId) messageId;
-        if (!emitted.contains(msgId)) {
-            if (msgId.isEmitted()) {
-                LOG.debug("Received ack for message [{}], associated with tuple emitted for a ConsumerRecord that "
-                    + "came from a topic-partition that this consumer group instance is no longer tracking "
-                    + "due to rebalance/partition reassignment. No action taken.", msgId);
+        if (isAtLeastOnceProcessing()) {
```

Comment:
I think this is worse. Putting a guard clause at the top decreases indentation for the whole function and is more readable IMO.

Reply:
I disagree. Any method that has only one if condition and nothing else, should be 
```java 
if (condition == true) 
     do_action;
```
imho it is counter natural to have code like 
```java
if (condition = false) 
    do_nothing;
else 
     do_action;
```
which is what basically the early return is doing.

There are also lengthier reasons related to the semantics of OOP, but I just think that in general wherever possible one should have code like if (condtion==true) do action.

Reply:
I know there's a huge debate on if else vs early return, but I'm in favor of early return, especially it is used as some pruning on early stage of logic.
Sure I'm not in favor of second statements (if do nothing else do action) but early return is completely different as do nothing, as it does pruning the logic branch and no need to follow the if statement and statements after end of if statement.

Reply:
Btw I don't think it is a big deal to decide and stick with specific approach, cause there's no statement after if statement. 
@srdo Could we leave the decision of this review comment to @hmcl?

Reply:
@hmcl I agree that 
```
if condition == false
  do nothing
else
  do something
```
is bad, but as @HeartSaVioR mentions early return (guard clause) isn't the same thing. Instead of writing
```
if precondition == true
  rest of function
return
```
it can make the function more readable to write 
```
if precondition == false
  return
rest of function
```
because the guarding `if` now doesn't cover the rest of the function, and the indendation level is reduced. See https://blog.codinghorror.com/flattening-arrow-code/ or https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting#Early_return.

I agree that since there's no code after the `if` closes it's not that important. Will leave it up to your judgement.